### PR TITLE
Per domain authorization for certificate issuance

### PR DESCRIPTION
**Pull request by charhate**
Introducing new Plugins `AuthorizationPlugin(Plugin)` and `DomainAuthorizationPlugin(AuthorizationPlugin)`. One can implement a `DomainAuthorizationPlugin` and configure its slug as `USER_DOMAIN_AUTHORIZATION_PROVIDER` to check if caller is authorized to issue a certificate for a given Common Name and Subject Alternative Name (SAN) of type DNSName.

Users authenticate to Lemur either using the API Key or using one of the authentication providers. Over the time, we have seen API keys being used by applications to talk to Lemur instead of direct user interaction. It seems logical to persist this information in Lemur for tracking purpose. Also it can be used for authorization. [e2d406ada25c_.py](lemur/migrations/versions/e2d406ada25c_.py) aims to add a new column `application_name` to `api_keys ` table. Null values are allowed making sure this change in backward compatible.

### CODE REVIEW

File: lemur/certificates/service.py [hunk 1]
```diff
@@ -1240,3 +1241,19 @@ def is_valid_owner(email):
...
+def allowed_issuance_for_domain(common_name, extensions):
+    check_permission_for_cn = True if common_name else False
+
+    # authorize issuance for every x509.DNSName SAN
+    if extensions and extensions.get("sub_alt_names"):
+        for san in extensions["sub_alt_names"]["names"]:
+            if isinstance(san, x509.DNSName):
+                if san.value == common_name:
+                    check_permission_for_cn = False
+                is_authorized_for_domain(san.value)
```
Comment (hosseinsh) [inline lemur/certificates/service.py:1255]: not returning the outcome of the AuthZ evaluation?

File: lemur/certificates/service.py [hunk 2]
```diff
@@ -1240,3 +1241,19 @@ def is_valid_owner(email):
...
+
+    # lemur UI copies CN as SAN (x509.DNSName). Permission check for CN might already be covered above.
+    if check_permission_for_cn:
+        is_authorized_for_domain(common_name)
```
Comment (hosseinsh) [inline lemur/certificates/service.py:1259]: not returning the outcome of the AuthZ evaluation?

File: lemur/certificates/views.py [hunk 3]
```diff
@@ -511,23 +512,21 @@ def post(self, data=None):
...
         authority_permission = AuthorityPermission(data["authority"].id, roles)
 
-        if authority_permission.can():
-            data["creator"] = g.user
+        if not authority_permission.can():
+            return dict(message=f"You are not authorized to use the authority: {data['authority'].name}"), 403
+
+        data["creator"] = g.user
+        try:
+            service.allowed_issuance_for_domain(data["common_name"], data["extensions"])
```
Comment (hosseinsh) [inline lemur/certificates/views.py:520]: same question as above about the outcome of the authZ eval

File: lemur/tests/test_certificates.py [hunk 4]
```diff
@@ -1578,3 +1578,86 @@ def start_server():
...
+    ],
+)
+def test_allowed_issuance_for_domain(common_name, extensions, expected_error, authz_check_count):
+    from lemur.certificates.service import allowed_issuance_for_domain
+
+    with patch(
+        'lemur.certificates.service.is_authorized_for_domain', side_effect=mocked_is_authorized_for_domain
+    ) as wrapper:
+        try:
+            allowed_issuance_for_domain(common_name, extensions)
```
Comment (hosseinsh) [inline lemur/tests/test_certificates.py:1654]: no return values being tested?

File: lemur/domains/service.py [hunk 5]
```diff
@@ -57,6 +60,30 @@ def is_domain_sensitive(name):
...
+
+    user_domain_authorization_provider = plugins.get(current_app.config.get("USER_DOMAIN_AUTHORIZATION_PROVIDER"))
+    caller = g.caller_application if hasattr(g, 'caller_application') else g.user.email
+    # if the caller can be mapped to an application name, use that to perform authorization
+    # this could be true when using API key to call lemur (migration script e2d406ada25c_.py)
+    _, error = user_domain_authorization_provider.is_authorized(domain=name, caller=caller)
+
+    if error:
+        raise error
+
```
[the remark this answers is not in the document]
Comment (charhate) [reply lemur/domains/service.py:87]: Regarding propagating of authz result, it instead relies on raising `UnauthorizedError`. Raising error with useful details within custom plugin implementation can be useful to return an actionable message when certificate issuance is denied. The exception will get caught at [certificate create Post call](https://github.com/Netflix/lemur/pull/3889/files#diff-88f1b5db302abda0bddba15f5af7d9c1aed03efcdd87d603bd0c3a96c41d772fR521) to return 403 along with the error message.
I did some minor changes to above method
1. look at `authorized` boolean, raise `UnauthorizedError()` if authz results False
2. update description to reflect behavior

File: lemur/domains/service.py [hunk 6]
```diff
@@ -57,6 +60,30 @@ def is_domain_sensitive(name):
...
+    domain.
+    Return True if authorized, False otherwise.
+    If authorization plugin is not available, return true by default
+
+    :param name: domain (string) for which authorization check is being done
+    :return: True/False
+    """
+    if current_app.config.get("USER_DOMAIN_AUTHORIZATION_PROVIDER") is None:
+        # nothing to check since USER_DOMAIN_AUTHORIZATION_PROVIDER is not configured
+        return
```
[the remark this answers is not in the document]
Comment (charhate) [reply lemur/domains/service.py:75]: Thanks for catching, modified description ([additional details](https://github.com/Netflix/lemur/pull/3889/files#r766258425))

File: lemur/certificates/service.py (hunk 1 above)
Comment (charhate) [reply lemur/certificates/service.py:1255]: It raises `UnauthorizedError` if authz fails ([additional details](https://github.com/Netflix/lemur/pull/3889/files#r766258425))

File: lemur/certificates/service.py (hunk 2 above)
Comment (charhate) [reply lemur/certificates/service.py:1259]: Same as above

File: lemur/certificates/views.py (hunk 3 above)
Comment (charhate) [reply lemur/certificates/views.py:520]: It raises `UnauthorizedError` if authz fails ([additional details](https://github.com/Netflix/lemur/pull/3889/files#r766258425)), which gets caught below to return 403 along with error message

File: lemur/tests/test_certificates.py (hunk 4 above)
Comment (charhate) [reply lemur/tests/test_certificates.py:1654]: asserting on exception below ([additional details](https://github.com/Netflix/lemur/pull/3889/files#r766258425))

**Review by hosseinsh — APPROVED** [review]
This is an awesome improvement to Lemur's AuthZ features.

**Pull request merged**